From fa6c7e44139e77847790962f5ec4146207a2feed Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Mon, 26 Dec 2022 13:30:54 +0200
Subject: [PATCH] [SW-66407] gaudi3: add barrier before polling on mem cpy status

After configuring and enabling the mem_cpy engine FW is polling on the
completion status.
since the first read of the polling loop is very close to the 'enable'
bit write, and we don't have read-write coherency, the first read may
return before the enable bit was actually written, and outdated status
would be used.
A write barrier is added before the polling cycle in order to make sure
the engine is enabled before polling on the status.

Change-Id: I75ef944bec517f092502d31dfa22226c5e0287ad
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index b60b0d2..b1da815 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -1149,6 +1149,9 @@
 
 	/* Enable copy and wait for completion */
 	regmap_writel(D0_PSOC_GLOBAL_CONF_MEM_CPY_CTRL_EN_M, mmD0_PSOC_GLOBAL_CONF_MEM_CPY_CTRL);
+	/* let 'enable' command propagate before polling on status */
+	HL_ARC_W_BARRIER();
+
 	while (1) {
 		cpy_sts = regmap_readl(mmD0_PSOC_GLOBAL_CONF_MEM_CPY_STATUS);
 		if (cpy_sts & D0_PSOC_GLOBAL_CONF_MEM_CPY_STATUS_DONE_M)
